compiler/src/dmd/nogc.d

```diff
@@ -138,9 +139,17 @@ public:
             return;                     // separate allocator is called for this, not the GC
         if (f.setGC())
         {
-            e.error("cannot use `new` in `@nogc` %s `%s`",
-                f.kind(), f.toPrettyChars());
-            err = true;
+            if (e.member && !(e.member.storage_class & STC.scope_) && e.type.isTypeClass())
+            {
+                deprecation(e.loc, "cannot use `new` in `@nogc` %s `%s`, perhaps annotate constructor for `class %s` with `scope`",
```

Comment:
I suggest
a) moving the deprecation out of the `@nogc` violation (so that e.g. the frontend regressions become visible), and
b) retaining the old behavior and instead deprecating the missing `scope` for the constructor, giving people time to adapt their codebases with the extra attribute *without* pessimizing currently working code.

Reply:
The trouble with this is it is perfectly legal to "scope-ize" any pointer to anything by giving it a storage class of `scope`. Storing the allocation on the stack is an optimization if the initializer meets additional criteria.

I.e. adding a deprecation for it will break existing code.